keps/sig-scheduling/4671-gang-scheduling/README.md

```diff
@@ -0,0 +1,983 @@
+
+The KEP must have a approver from the
+[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
+team. Please reach out on the
+[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
+you need any help or guidance.
+-->
+
+### Feature Enablement and Rollback
+
+<!--
+This section must be completed when targeting alpha to a release.
+-->
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [X] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: Workload/GenericWorkload/NativeWorkload
+  - Components depending on the feature gate:
+    - kube-apiserver
+    - kube-scheduler
+  - Feature gate name: GangScheduling
+  - Components depending on the feature gate:
+    - kube-scheduler
+- [ ] Other
+  - Describe the mechanism:
+  - Will enabling / disabling the feature require downtime of the control
+    plane?
+  - Will enabling / disabling the feature require downtime or reprovisioning
+    of a node?
+
+###### Does enabling the feature change any default behavior?
+
+No. Gang scheduling is triggerred purely via existence of Workload objects and
+those are not yet created automatically behind the scenes.
+
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes. The GangScheduling features gate need to be switched off to disabled gang scheduling
+functionality.
+If additionally the API changes needs to be disabled, the Workload feature gate needs to
+also be disabled. However, the content of `spec.workload` fields in Pod objects will not be
+cleared, as well as the existing Workload objects will not be deleted.
+
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+The feature should start working again.
+However, the user need to remember that some Workload objects could already be stored
+in etcd and may affect the behavior of some of the existing workloads.
+
+
+###### Are there any tests for feature enablement/disablement?
+
+No.
+The enablement/disablement for the new field in Pod API will be added similarly to this PR:
+https://github.com/kubernetes/kubernetes/pull/97058/files#diff-7826f7adbc1996a05ab52e3f5f02429e94b68ce6bce0dc534d1be636154fded3R246-R282
+
+Note that gang-scheduling itself is purely in-memory feature, so feature themselves are enough.
+
+
+### Rollout, Upgrade and Rollback Planning
+
+<!--
+This section must be completed when targeting beta to a release.
+-->
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+<!--
+Try to be as paranoid as possible - e.g., what if some components will restart
+mid-rollout?
+
+Be sure to consider highly-available clusters, where, for example,
+feature flags will be enabled on some API servers and not others during the
+rollout. Similarly, consider large clusters and how enablement/disablement
+will rollout across nodes.
+-->
+
+###### What specific metrics should inform a rollback?
```

Comment:
Although not required, have you considered what kind of metrics you could expose for this functionality and where? I'm thinking about exposing gang-related metrics in the scheduler, as the most appropriate place to track this.

Reply:
I was thinking primarily about the gangscheduling feature, since like most of our scheduling code it's hard to follow from outside. I like both of the proposed metrics. Now I need to only remember that you wrote them here for beta :sweat_smile: